pkg/proxy/endpointslicecache.go

```diff
@@ -290,15 +290,13 @@ func (cache *EndpointSliceCache) addEndpointsByIP(serviceNN types.NamespacedName
 		endpointInfo := newBaseEndpointInfo(endpoint.Addresses[0], portNum, isLocal, endpoint.Topology,
 			endpoint.Ready, endpoint.Serving, endpoint.Terminating, endpoint.ZoneHints)
 
-		// This logic ensures we're deduping potential overlapping endpoints
-		// isLocal should not vary between matching IPs, but if it does, we
-		// favor a true value here if it exists.
-		if _, exists := endpointsByIP[endpointInfo.IP()]; !exists || isLocal {
-			endpointsByIP[endpointInfo.IP()] = cache.makeEndpointInfo(endpointInfo)
+		// This logic ensures we're deduplicating potential overlapping endpoints
+		if _, exists := endpointSet[endpointInfo.String()]; !exists {
+			endpointSet[endpointInfo.String()] = cache.makeEndpointInfo(endpointInfo)
```

Comment:
This is a great fix and should actually result in a bit of a performance improvement as well.